internal/tofu/write_only.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tofu
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/opentofu/opentofu/internal/configs"
+	"github.com/opentofu/opentofu/internal/configs/configschema"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// validateWriteOnlyAttributes checks that the write-only attributes are not returned back with an actual value.
+func validateWriteOnlyAttributes(schema *configschema.Block, v cty.Value, resCfg *configs.Resource) (diags tfdiags.Diagnostics) {
```

Comment:
I have mixed feelings about this.  It's a valid solution, but leaks the "provider should be implemented correctly" concerns outside the providers.Interface.

What do you think about implementing this logic within the objects that fulfill providers.Interface instead?

Reply:
Thanks! I struggled with finding a good place for this and I don't like it either the way it is now. Let me try and fix this.

Reply:
Fixed in 906d5e6e704de6328995140ba854cb849086fc49.
The issue with this approach is that at the Interface level we have no way to reference back the resource configuration as was done before.
But that shouldn't be a problem since the issue that it's validated here is not per resource instance but it's checking the required functionality of the whole provider instance. We should be fine.

Reply:
Thanks!